torch/distributions/wishart.py

```diff
@@ -0,0 +1,198 @@
+
+        if scale_tril is not None:
+            self._unbroadcasted_scale_tril = scale_tril
+        elif covariance_matrix is not None:
+            self._unbroadcasted_scale_tril = torch.linalg.cholesky(covariance_matrix)
+        else:  # precision_matrix is not None
+            self._unbroadcasted_scale_tril = _precision_to_scale_tril(precision_matrix)
+
+        self.dist_gamma = torch.distributions.Gamma(
+            self.df - torch.arange(self._event_shape[0], device=self._unbroadcasted_scale_tril.device) / 2,
+            1 / 2,
+        )
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(Wishart, _instance)
+        batch_shape = torch.Size(batch_shape)
+        cov_shape = batch_shape + self.event_shape
+        new._unbroadcasted_scale_tril = self._unbroadcasted_scale_tril
+        if 'covariance_matrix' in self.__dict__:
+            new.covariance_matrix = self.covariance_matrix.expand(cov_shape)
+        if 'scale_tril' in self.__dict__:
+            new.scale_tril = self.scale_tril.expand(cov_shape)
+        if 'precision_matrix' in self.__dict__:
+            new.precision_matrix = self.precision_matrix.expand(cov_shape)
+        super(Wishart, new).__init__(batch_shape,
+                                     self.event_shape,
+                                     validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    @lazy_property
+    def scale_tril(self):
+        return self._unbroadcasted_scale_tril.expand(
+            self._batch_shape + self._event_shape)
+
+    @lazy_property
+    def covariance_matrix(self):
+        return torch.einsum(
+            "ik,jk->ij",
+            self._unbroadcasted_scale_tril,
+            self._unbroadcasted_scale_tril
+        ).expand(self._batch_shape + self._event_shape)
+
+    @lazy_property
+    def precision_matrix(self):
+        identity = torch.eye(
+            self._event_shape[0],
+            device=self._unbroadcasted_scale_tril.device,
+            dtype=self._unbroadcasted_scale_tril.dtype,
+        )
+        return torch.cholesky_solve(
+            identity, self._unbroadcasted_scale_tril
+        ).expand(self._batch_shape + self._event_shape)
+
+    @property
+    def mean(self):
+        return self.df * self.covariance_matrix
+
+    @property
+    def variance(self):
+        V = self.covariance_matrix
+        diag_V = V.diag()
+        return self.df * (V.pow(2) + torch.einsum("i,j->ij", diag_V, diag_V))
+
+    def rsample(self, sample_shape=torch.Size()):
+        shape = self._extended_shape(sample_shape)
+
+        # Implemented Bartlett decomposition
+        if self.bartlett_decomposition:
+            noise = self.dist_gamma.sample(sample_shape).diag_embed(dim1=-2, dim2=-1).sqrt()
+            noise = noise + torch.randn(shape, device=noise.device).tril(diagonal=-1)
+        else:
+            noise = _standard_normal(
```

Comment:
This seems very different from the `noise` above, could you add a comment describing this block?

Reply:
Thanks for checking! There is an error in implementation.
I tried to implement original, standard implementation of Wishart sampling. Which requires 'df' parameter to be a positive integer.
However, in Bartlett decomposition, it is required that 'df' parameter to be a positive real, not positive interger.
I will figure out the better practice to consider both cases in the '.rsample' method.

Reply:
I like the latter idea - we can probably remove this and use Bartlett by default.

Reply:
Okay. I will use the Bartlett decomposition only.